tools/bug-report/pkg/bugreport/bugreport.go

```diff
@@ -174,6 +179,67 @@ func runBugReportCommand(_ *cobra.Command, logOpts *log.Options) error {
 	return nil
 }
 
+func dumpRevisionsAndVersions(resources *cluster2.Resources, kubeconfig, configContext, istioNamespace string) {
+	text := ""
+	revisions := getIstioRevisions(resources)
+	istioVersions, proxyVersions := getIstioVersions(kubeconfig, configContext, istioNamespace, revisions)
+	text += "The following Istio control plane revisions/versions were found in the cluster:\n"
+	for rev, ver := range istioVersions {
+		text += fmt.Sprintf("Revision %s:\n%s\n\n", rev, ver)
+	}
+	text += "The following proxy revisions/versions were found in the cluster:\n"
+	for rev, ver := range proxyVersions {
+		text += fmt.Sprintf("Revision %s: Versions {%s}\n", rev, strings.Join(ver, ", "))
+	}
+	common.LogAndPrintf(text)
+	writeFile(filepath.Join(archive.OutputRootDir(tempDir), "versions"), text)
+}
+
+// getIstioRevisions returns a slice with all Istio revisions detected in the cluster.
+func getIstioRevisions(resources *cluster2.Resources) []string {
+	var out []string
+	for _, podLabels := range resources.Labels {
+		for label, value := range podLabels {
+			if label == istioRevisionLabel {
+				out = append(out, value)
+			}
+		}
+	}
+	return out
+}
+
+// getIstioVersions returns a mapping of revision to aggregated version string for Istio components and revision to
+// slice of versions for proxies. Any errors are embedded in the revision strings.
+func getIstioVersions(kubeconfig, configContext, istioNamespace string, revisions []string) (map[string]string, map[string][]string) {
+	istioVersions := make(map[string]string)
+	proxyVersions := make(map[string][]string)
+	for _, revision := range revisions {
```

Comment:
won't this list end up with possible many duplicate labels, or (MANY labels, if it includes sidecars, which I think it does)? Should revisions be deduped?

Reply:
good catch, done.

Reply:
you are still calling `getIstioVersion` and `GetProxyInfo` which is expensive for each one I think.

Reply:
yes, but they're called just once per revision - i don't see a way to call them fewer times than that.